src/garage/torch/algos/vpg.py

```diff
@@ -0,0 +1,286 @@
+        self.entropy_regularzied = (entropy_method == 'regularized')
+        #     Entropy regularization, max entropy rl
+
+        self._check_entropy_configuration(entropy_method, center_adv,
+                                          stop_entropy_gradient,
+                                          policy_ent_coeff)
+
+        self.reward_mean = collections.deque(maxlen=100)
+
+        self.optimizer = optimizer(
+            self.policy.get_module().parameters(), lr=self.policy_lr)
+
+    def _check_entropy_configuration(self, entropy_method, center_adv,
+                                     stop_entropy_gradient, policy_ent_coeff):
+        if entropy_method not in ('max', 'regularized', 'no_entropy'):
+            raise ValueError('Invalid entropy_method')
+
+        if entropy_method == 'max':
+            if center_adv:
+                raise ValueError('center_adv should be False when '
+                                 'entropy_method is max')
+            if not stop_entropy_gradient:
+                raise ValueError('stop_gradient should be True when '
+                                 'entropy_method is max')
+        if entropy_method == 'no_entropy':
+            if policy_ent_coeff != 0.0:
+                raise ValueError('policy_ent_coeff should be zero '
+                                 'when there is no entropy method')
+
+    def train_once(self, itr, paths):
+        """Perform one step of policy optimization."""
+        self._optimize_policy(itr, paths)
+        self._optimize_baseline(itr, paths)
+
+        return np.mean([sum(path['rewards']) for path in paths])
+
+    def _optimize_policy(self, itr, paths):
+        loss = [self._get_policy_loss(path) for path in paths]
+
+        self.optimizer.zero_grad()
+        torch.cat(loss).mean().backward()
+        self.optimizer.step()
+
+    def _get_policy_loss(self, path):
+        rewards = torch.Tensor(path['rewards'])
+        obs = torch.Tensor(path['observations'])
+
+        policy_entropy = self._get_policy_entropy(obs)
+
+        if self.maximum_entropy:
+            rewards = [
+                reward + self.policy_ent_coeff * policy_entropy
+                for reward in rewards
+            ]
+
+        baselines = self._get_baselines(path)
+        path['baselines'] = baselines.detach().numpy()
+        advantages = self._compute_advantages(baselines, rewards)
+
+        if self.center_adv:
+            advantages = F.batch_norm(
+                advantages.reshape((-1, 1)),
+                advantages.mean().unsqueeze(0),
+                advantages.var().unsqueeze(0),
+                eps=self.eps)
+            advantages = advantages.squeeze()
+
+        if self.positive_adv:
+            advantages -= advantages.min()
+
+        log_likelihood = self.policy.get_log_likelihood(
+            torch.Tensor(path['observations']), torch.Tensor(path['actions']))
+
+        grad = (log_likelihood * advantages).sum()
```

Comment:
Hmm this does not looks very correct to me. I think what you want is actually take the mean across the whole batch so this should not be a sum. Having that said, I think you should firstly flatten the batch dimension before calculating the loss instead of just pass each path to calculate the loss. Though, they are very similar, but not 100% the same in some cases.

Reply:
Also let's just rename this variable to something else other than grad? This is not really gradients.

Reply:
![image](https://user-images.githubusercontent.com/43084978/62735172-726b7000-b9df-11e9-8b83-0841b99d2d53.png)

I thought we need to sum the (log likelihood * advantage) and then get mean value across the sum values.

Could you explain more about this part?

Reply:
I think the `tf.dynamic_partition` has an important role in "[validating](https://github.com/rlworkgroup/garage/blob/1d649d61b52830fd8303b52d0a3430af94e5b92b/src/garage/tf/misc/tensor_utils.py#L62)" true values from paddings. However, I cannot find PyTorch function similar to `tf.dynamic_partition`.

Also, for [computing advantage value](https://github.com/rlworkgroup/garage/blob/1d649d61b52830fd8303b52d0a3430af94e5b92b/src/garage/tf/misc/tensor_utils.py#L250), (I am not sure about this but,) TensorFlow Conv1d is different from PyTorch Conv1d. (PyTorch doesn't have a filter.. (instead it has 'kernel size').
And, I can't use discount factor in PyTorch cumsum or cumprod.

Do you have any idea?

Reply:
The `D_k` here should be referring to the sampled batch so this is divided the size of the whole batch. You can see it as taking a mean across the whole batch. Please verify since I did not take a look at the paper but I believe it is the right thing to do.

Reply:
I think you can actually just do the validation by multiply everything with the `valids`.

Reply:
@yonghyuc I agree with zhanpenghe. At some point, you'll need to divide the reward/-loss by the number of trajectories provided by the sampler. Because the gradient is a linear operation, you can do that here on line 162.